src/manager/checkpoint.rs

```diff
@@ -164,87 +164,103 @@ async fn manage_subnet((child, parent): (Subnet, Subnet), stop_notify: Arc<Notif
             let subnet_actor_state = parent_client
                 .ipc_read_subnet_actor_state(&child.id, parent_tip_set)
                 .await?;
-
-            let mut validator_set: HashSet<Address, RandomState> = HashSet::new();
-            match subnet_actor_state.validator_set.validators {
-                None => {}
-                Some(validators) => {
-                    for v in validators {
-                        validator_set.insert(Address::from_str(v.addr.deref())?);
+            let last_exec = subnet_actor_state
+                .bottom_up_checkpoint_voting
+                .last_voting_executed;
+            let submission_epoch = last_exec + period;
+
+            if curr_epoch >= submission_epoch {
+                // First, we check which accounts are in the validator set. This is done by reading
+                // the parent's chain head and requesting the state at that tip set.
+                let parent_head = parent_client.chain_head().await?;
+                // A key assumption we make now is that each block has exactly one tip set. We panic
+                // if this is not the case as it violates our assumption.
+                // TODO: update this logic once the assumption changes (i.e., mainnet)
+                assert_eq!(parent_head.cids.len(), 1);
+                let cid_map = parent_head.cids.first().unwrap().clone();
+                let parent_tip_set = Cid::try_from(cid_map)?;
+
+                let mut validator_set: HashSet<Address, RandomState> = HashSet::new();
+                match subnet_actor_state.validator_set.validators {
+                    None => {}
+                    Some(validators) => {
+                        for v in validators {
+                            validator_set.insert(Address::from_str(v.addr.deref())?);
+                        }
+                    }
+                };
+
+                // Now, for each account defined in the `child` subnet that is in the validator set, we
+                // submit a checkpoint on its behalf.
+                assert_eq!(child_head.cids.len(), 1); // Again, check key assumption
+                let child_tip_set = Cid::try_from(child_head.cids.first().unwrap().clone())?;
+                for account in child.accounts.iter() {
+                    if validator_set.contains(account) {
+                        // check if the validator already voted
+                        let has_voted = parent_client
+                            .ipc_validator_has_voted_bottomup(&child.id, epoch, account)
+                            .await
+                            .map_err(|e| {
+                                log::error!(
+                                    "error checking if validator has voted in subnet: {:?}",
+                                    &child.id
+                                );
+                                e
+                            })?;
+                        if !has_voted {
```

Comment:
Instead of this, maybe we can take the `child.accounts` vector and filter it first in a functional way for the accounts that are in the validator set and haven't voted.

Reply:
Let's discuss this one at some point, I may not be proficient enough in Rust, but I can't seem to be able to make  async blocks work inside `filter`. Once I have a bit more time to breath I will dig a bit more

Reply:
No worries. It was just a nitpick suggestion.